prov/sm2/src/sm2_atomic.c

```diff
@@ -0,0 +1,534 @@
+		}
+		break;
+	default:
+		assert(0);
+		break;
+	}
+
+	ofi_spin_lock(&ep->tx_lock);
+
+	ret = sm2_do_atomic_inject(
+		ep, peer_gid, op, op_flags, datatype, atomic_op, iov, iov_count,
+		rma_ioc, rma_ioc_count, compare_iov, compare_count, result_iov,
+		result_count, total_len, context);
+
+	if (ret)
+		goto out;
+
+	if (!(op_flags & FI_REMOTE_READ) &&
+	    !(op_flags & FI_DELIVERY_COMPLETE)) {
+		ret = sm2_complete_tx(ep, context, op, op_flags);
+		if (ret) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"unable to process tx completion\n");
+		}
+	}
+
+out:
+	ofi_spin_unlock(&ep->tx_lock);
+	return ret;
+}
+
+static ssize_t sm2_atomic_writemsg(struct fid_ep *ep_fid,
+				   const struct fi_msg_atomic *msg,
+				   uint64_t flags)
+{
+	struct sm2_ep *ep;
+
+	ep = container_of(ep_fid, struct sm2_ep, util_ep.ep_fid.fid);
+
+	return sm2_generic_atomic(
+		ep, msg->msg_iov, msg->desc, msg->iov_count, NULL, NULL, 0,
+		NULL, NULL, 0, msg->addr, msg->rma_iov, msg->rma_iov_count,
+		msg->datatype, msg->op, msg->context, ofi_op_atomic,
+		flags | ep->util_ep.tx_msg_flags);
+}
+
+static ssize_t sm2_atomic_writev(struct fid_ep *ep_fid,
+				 const struct fi_ioc *iov, void **desc,
+				 size_t count, fi_addr_t dest_addr,
+				 uint64_t addr, uint64_t key,
+				 enum fi_datatype datatype, enum fi_op op,
+				 void *context)
+{
+	struct sm2_ep *ep;
+	struct fi_rma_ioc rma_iov;
+
+	ep = container_of(ep_fid, struct sm2_ep, util_ep.ep_fid.fid);
+
+	rma_iov.addr = addr;
+	rma_iov.count = ofi_total_ioc_cnt(iov, count);
+	rma_iov.key = key;
+
+	return sm2_generic_atomic(ep, iov, desc, count, NULL, NULL, 0, NULL,
+				  NULL, 0, dest_addr, &rma_iov, 1, datatype, op,
+				  context, ofi_op_atomic, sm2_ep_tx_flags(ep));
+}
+
+static ssize_t sm2_atomic_write(struct fid_ep *ep_fid, const void *buf,
+				size_t count, void *desc, fi_addr_t dest_addr,
+				uint64_t addr, uint64_t key,
+				enum fi_datatype datatype, enum fi_op op,
+				void *context)
+{
+	struct sm2_ep *ep;
+	struct fi_ioc iov;
+	struct fi_rma_ioc rma_iov;
+
+	ep = container_of(ep_fid, struct sm2_ep, util_ep.ep_fid.fid);
+
+	iov.addr = (void *) buf;
+	iov.count = count;
+
+	rma_iov.addr = addr;
+	rma_iov.count = count;
+	rma_iov.key = key;
+
+	return sm2_generic_atomic(ep, &iov, &desc, 1, NULL, NULL, 0, NULL, NULL,
+				  0, dest_addr, &rma_iov, 1, datatype, op,
+				  context, ofi_op_atomic, sm2_ep_tx_flags(ep));
+}
+
+static ssize_t sm2_atomic_inject(struct fid_ep *ep_fid, const void *buf,
+				 size_t count, fi_addr_t dest_addr,
+				 uint64_t addr, uint64_t key,
+				 enum fi_datatype datatype,
+				 enum fi_op atomic_op)
+{
+	struct sm2_ep *ep =
+		container_of(ep_fid, struct sm2_ep, util_ep.ep_fid.fid);
+	struct iovec iov;
+	struct fi_rma_ioc rma_ioc;
+	sm2_gid_t peer_gid;
+	ssize_t ret = 0;
+	size_t total_len;
+
+	ret = sm2_verify_peer(ep, dest_addr, &peer_gid);
+	if (ret < 0)
+		return ret;
+
+	total_len = count * ofi_datatype_size(datatype);
+	assert(total_len <= SM2_ATOMIC_INJECT_SIZE);
```

Comment:
This assert is the only place I could find which ensures user does not request atomics on a buffer which is too large.

I think we need to do a check on size and return EINVAL or similar if the user attempts to send too many atomics at once.  This assert will do nothing in non-debug code.

Reply:
I think all of the asserts are ok as asserts. The ofi_copy_to/from functions take in the maximum amount of data to copy (inject size) so we shouldn't overwrite the data. Other providers treat errors like this similarly (for example in efa [here](https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/rdm/efa_rdm_msg.c#LL202)). I don't think it's worth trying to cleanly catch any API misuse errors the application has since it hurts the fast path.

Reply:
Ah, I missed the maximum amount to copy.  Mostly I wanted to make sure we didn't overrun some buffer because application was negligent.  OK, looks good to me.